### PR TITLE
SONAR-9352 do not let quality gate fail for very small changesets

### PR DESCRIPTION
**Pull request by bartfastiel**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  private final MeasureRepository measureRepository;
+  private final MetricRepository metricRepository;
+
+  public SmallChangesetQualityGateSpecialCase(MeasureRepository measureRepository, MetricRepository metricRepository) {
+    this.measureRepository = measureRepository;
+    this.metricRepository = metricRepository;
+  }
+
+  QualityGateMeasuresStep.MetricEvaluationResult applyIfNeeded(Component project, @Nullable QualityGateMeasuresStep.MetricEvaluationResult metricEvaluationResult) {
+    if (metricEvaluationResult.evaluationResult == null) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: MAJOR A "NullPointerException" could be thrown; "metricEvaluationResult" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+public class SmallChangesetQualityGateSpecialCase {
+
+  /**
+   * Some metrics will only produce warnings (never errors) on very small change sets.
+   */
+  private static final Collection<String> METRICS_THAT_CAN_ONLY_PRODUCE_WARNINGS_ON_SMALL_CHANGESETS = asList(
+    CoreMetrics.COVERAGE_KEY, CoreMetrics.NEW_COVERAGE_KEY,
+    CoreMetrics.LINE_COVERAGE_KEY, CoreMetrics.NEW_LINE_COVERAGE_KEY);
+  //FIXME add more metrics
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    if (!METRICS_THAT_CAN_ONLY_PRODUCE_WARNINGS_ON_SMALL_CHANGESETS.contains(metricEvaluationResult.condition.getMetric().getKey())) {
+      return metricEvaluationResult;
+    }
+    if (!isSmallChangeset(project)) {
+      return metricEvaluationResult;
+    }
+    return calculateModifiedResult(metricEvaluationResult);
+  }
+
+  private QualityGateMeasuresStep.MetricEvaluationResult calculateModifiedResult(@Nullable QualityGateMeasuresStep.MetricEvaluationResult metricEvaluationResult) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: MAJOR Make "calculateModifiedResult" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    return new QualityGateMeasuresStep.MetricEvaluationResult(
+      new EvaluationResult(Measure.Level.OK, metricEvaluationResult.evaluationResult.getValue()), metricEvaluationResult.condition);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: MAJOR A "NullPointerException" could be thrown; "metricEvaluationResult" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    if (metricEvaluationResult == null) {
+      return false;
+    }
+    if (metricEvaluationResult.evaluationResult.getLevel() == Measure.Level.OK) {
+      return false;
+    }
+    if (!METRICS_THAT_CAN_ONLY_PRODUCE_WARNINGS_ON_SMALL_CHANGESETS.contains(metricEvaluationResult.condition.getMetric().getKey())) {
+      return false;
+    }
+    if (!isSmallChangeset(project)) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: MINOR Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    if (metricEvaluationResult == null) {
+      return false;
+    }
+    if (metricEvaluationResult.evaluationResult.getLevel() == Measure.Level.OK) {
+      return false;
+    }
+    if (!METRICS_TO_IGNORE_ON_SMALL_CHANGESETS.contains(metricEvaluationResult.condition.getMetric().getKey())) {
+      return false;
+    }
+    if (!isSmallChangeset(project)) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: MINOR Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

**Comment by stas-vilchik** [comment]
@gregaubert please review my commit here

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  private final MeasureRepository measureRepository;
+  private final MetricRepository metricRepository;
+
+  public SmallChangesetQualityGateSpecialCase(MeasureRepository measureRepository, MetricRepository metricRepository) {
+    this.measureRepository = measureRepository;
+    this.metricRepository = metricRepository;
+  }
+
+  QualityGateMeasuresStep.MetricEvaluationResult applyIfNeeded(Component project, @Nullable QualityGateMeasuresStep.MetricEvaluationResult metricEvaluationResult) {
+    if (metricEvaluationResult == null) {
```
[the remark this answers is not in the document]
Comment (teryk) [reply server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java]: can be simplified

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGateDetailsFormatter.java [hunk 8]
```diff
@@ -55,6 +55,13 @@ public QualityGateDetailsFormatter(Optional<String> measureData, Optional<Snapsh
     ProjectStatusWsResponse.Status qualityGateStatus = measureLevelToQualityGateStatus(json.get("level").getAsString());
     projectStatusBuilder.setStatus(qualityGateStatus);
 
+    JsonElement ignoredConditions = json.get("ignoredConditions");
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGateDetailsFormatter.java]: extract in method `formatIgnoredConditions`

**Comment by bartfastiel** [comment]
I'm waiting for the PM review before I merge.

**Pull request merged**